overlord/configstate/transaction.go

```diff
@@ -0,0 +1,172 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configstate
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// Transaction is responsible for configuration transactions. Its config will
+// never change outside the control of its owner, and any changes made to it
+// doesn't affect the overall system config. Change it, query it, and if you
+// really want to get any changes saved back into the config, Commit it.
+//
+// Note that Transactions are safe to access concurrently.
+type Transaction struct {
+	state           *state.State
+	config          systemConfig
+	configMutex     sync.RWMutex
+	writeCache      systemConfig
+	writeCacheMutex sync.RWMutex
+}
+
+type snapConfig map[string]*json.RawMessage
+type systemConfig map[string]snapConfig
+
+// NewTransaction creates a new config transaction initialized with the given state.
```

Comment:
Yeah, in this case having the lock out enables the transaction to be created while other things are also done in the same state transaction. So it's the right thing to do.